src/main/scala/org/apache/spark/shuffle/ucx/UcxShuffleTransport.scala

```diff
@@ -0,0 +1,229 @@
+/*
+* Copyright (c) 2020, NVIDIA CORPORATION. ALL RIGHTS RESERVED.
+* See file LICENSE for terms.
+*/
+package org.apache.spark.shuffle.ucx
+import java.io.ObjectInputStream
+import java.nio.ByteBuffer
+import java.util.concurrent.ConcurrentHashMap
+
+import com.fasterxml.jackson.databind.util.ByteBufferBackedInputStream
+import org.openucx.jucx.UcxUtils
+import org.openucx.jucx.ucp._
+import org.apache.spark.SparkEnv
+import org.apache.spark.internal.Logging
+import org.apache.spark.shuffle.ucx.memory.{MemoryPool, UcxHostBounceBuffersPool}
+import org.apache.spark.shuffle.ucx.rpc.GlobalWorkerRpcThread
+import org.apache.spark.util.Utils
+
+/**
+ * Special type of [[ Block ]] interface backed by UcpMemory,
+ * it may not be actually pinned if used with [[ UcxShuffleConf.useOdp ]] flag.
+ */
+case class UcxPinnedBlock(block: Block, ucpMemory: UcpMemory) extends Block {
+  override def getMemoryBlock: MemoryBlock = MemoryBlock(ucpMemory.getAddress, ucpMemory.getLength)
+}
+
+class UcxStats extends OperationStats {
+  private[ucx] val startTime = System.nanoTime()
+  private[ucx] var endTime: Long = 0L
+  private[ucx] var receiveSize: Long = 0L
+
+  /**
+   * Time it took from operation submit to callback call.
+   * This depends on [[ ShuffleTransport.progress() ]] calls,
+   * and does not indicate actual data transfer time.
+   */
+  override def getElapsedTimeNs: Long = endTime - startTime
+
+  /**
+   * Indicates number of valid bytes in receive memory when using
+   * [[ ShuffleTransport.fetchBlockByBlockId()]]
+   */
+  override def recvSize: Long = receiveSize
+}
+
+class UcxRequest(request: UcpRequest, stats: OperationStats) extends Request {
+
+  override def isCompleted: Boolean = request.isCompleted
+
+  override def cancel(): Unit = request.close()
+
+  override def getStats: Option[OperationStats] = Some(stats)
+}
+
+/**
+ * UCX implementation of [[ ShuffleTransport ]] API
+ */
+class UcxShuffleTransport(var ucxShuffleConf: UcxShuffleConf = null)
+  extends ShuffleTransport with Logging {
+  // UCX entities
+  private var ucxContext: UcpContext = _
+  private var globalWorker: UcpWorker = _
+  private val ucpWorkerParams = new UcpWorkerParams()
+
+  // TODO: reimplement as workerPool, since spark may create/destroy threads dynamically
+  private var threadLocalWorker: ThreadLocal[UcxWorkerWrapper] = _
+  private val allocatedWorkers = ConcurrentHashMap.newKeySet[UcxWorkerWrapper]
+  private var progressThread: Thread = _
+
+  private val registeredBlocks: ConcurrentHashMap[BlockId, Block] = new ConcurrentHashMap
+  private val memMapParams = new UcpMemMapParams()
+
+  if (ucxShuffleConf.useOdp) {
+    memMapParams.nonBlocking()
+  }
+
+  //Mapping between executorId and it's address
+  private[ucx] val executorIdToAddress = new ConcurrentHashMap[String, ByteBuffer]()
+  private[ucx] val clientConnections = new ConcurrentHashMap[ByteBuffer, UcpEndpoint]()
+
+  var memoryPool: MemoryPool = _
```

Comment:
Would be good to know what the use cases are for `memoryPool`, Could you add a comment for it?